File: java/flesch.java
import java.util.*;
import java.io.*;

public class flesch
{
    public static void main(String[] args) throws FileNotFoundException {

        double alpha, beta, index, gindex = 0;  // instantiate necessary variables

        String filename;                        // holds filename

        // Create scanner that reads in the file, open 
        // file either from command line or from user input
        Scanner in = new Scanner(System.in);
        if (args.length == 0){
            System.out.print("Please enter a file name (must exist in current directory): ");
            filename = in.nextLine();
        }
        else{
            filename = args[0];
        }

        System.out.print("\nFile being used: ");
        System.out.print(filename);
        System.out.print("\n");

        ArrayList<String> wordArr = new ArrayList<String>(50);      // set up arraylist for all words
        ArrayList<String> sentArr = new ArrayList<String>(50);      // Set up arraylist used to count sentences

        alpha = getAlpha(filename, wordArr);                // Calculate alpha
        beta = countWords(filename, wordArr) / countSentences(filename, sentArr); // Calculate beta

        index = 206.835 - alpha * 84.6 - beta * 1.015;      // Calculate index
        gindex = alpha * 11.8 + beta * 0.39 - 15.59;        // Calculate grade index

        // Print the stuff
        System.out.print("Index = ");
        System.out.print(Math.round(index));
        System.out.print("\n");
        System.out.print("Grade Level Index = ");
        System.out.print((Math.round(gindex)  * 10) / 10.0);
        System.out.print("\n\n");
    }

    public static boolean isAlphabetic(String check){
        /*
         *  checks string to see if it's alphabetic
         */
        char[] chars = check.toCharArray();             // turn string into char array 
        for (char c : chars){                           // for all chars in the string
            if(!Character.isLetter(c)) {                // if curr is not a letter, then
                return false;                           // return false
            }
        }
        return true;
    }

    public static boolean charIsAlphabetic(char check){
        /*
         *  checks char to see if it's alphabetic
         */
        if(!Character.isLetter(check))
            return false;
        return true;    

    }

    public static boolean isVowel(char check){
        /*
         *  Checks char to see if it's a vowel
         */
        if (check ==  'a' || check ==  'e' || check ==  'i' || check ==  'o' || check ==  'u' || check ==  'y' || check ==  'A' || check ==  'E' || check ==  'I' || check ==  'O' || check ==  'U' || check ==  'Y'){
            return true;
        }
        return false;

    }

    public static int countSyllables(String word) {
    /*
     *  Count all syllables within the arraylist of words
     */
        String input = word.toLowerCase(); // convert string to lower case
        int syllables = 0, eSatEnd=0;       

        int i = input.length() - 1;         // set i to length-1 to not go out of bounds

        // count all the e's in the end
        while (i >= 0 && input.charAt(i) == 'e') {
            i--;
            eSatEnd++;
        }
        
        // if there's an e at the end, there's automatically one syllable in the word
        if (eSatEnd == 1) {
            syllables = 1;
        }

        boolean vowelBefore = false;        // holds bool of if vowel is adjacent to curr vowel
        while (i >= 0) {
            if (isVowel(input.charAt(i))) { // if there's a vowel here
                if (!vowelBefore) {         // and if no vowel before, 
                    syllables++;            // increment
                    vowelBefore = true;    
                }
            } else {
                vowelBefore = false;       
            }
            i--;
        }
        return syllables;
    }



    public static double countSentences(String filename, ArrayList<String> sentArr){
        /*
         *  Count all sentences within the infile using end-of-sentence markers
         */

        Scanner infile = null;
        try{
            infile = new Scanner(new File(filename));
        }
        catch (FileNotFoundException ex){
            System.out.println("File Not Found.");
            System.exit(0);
        }	
        double count = 0;

        while (infile.hasNextLine()){                       // While there's no EOF
            Scanner line = new Scanner(infile.nextLine()); 
            line.useDelimiter("");
            while(line.hasNext()){
                String ch = line.next();                    // get next string in infile's current line
                if (ch.contains(".") || ch.contains("!") || ch.contains("?") || ch.contains(";") || ch.contains(":")){
                    count++;
                }
            }
        }
        return count;
    }

    public static boolean isNum(String str){
        /*
         *  determine if str contains a number
         */
        
        if ((str.indexOf('0') != -1 || str.indexOf('1') != -1 || str.indexOf('2') != -1 || str.indexOf('3') != -1 || str.indexOf('4') != -1 || str.indexOf('5') != -1 || str.indexOf('6') != -1 || str.indexOf('7') != -1 || str.indexOf('8') != -1 || str.indexOf('9') != -1) && !isAlphabetic(str))
        {
            return true;
        }

        return false;
    }

    public static double getAlpha(String filename, ArrayList<String> wordArr){
        /*
         *  returns the alpha portion of the index and grade index calculation
         */
    
        double count = 0, alpha = 0, syllables = 0, sCount = 0;
        Scanner infile = null;
        try{
            infile = new Scanner(new File(filename));  
        }
        catch (FileNotFoundException ex){
            System.out.println("File Not Found.");
            System.exit(0);
        }
        String temp;
        while(infile.hasNextLine()){
            Scanner line = new Scanner(infile.nextLine());
            while(line.hasNext()){
                wordArr.add(line.next());               // add curr string to the array of words
                count++;
            }
        }
        for(int i = 0; i < wordArr.size(); i++){
            syllables+=((double)countSyllables(wordArr.get(i)));    // add the int returned from the syllables method to the syllable count
        }
        alpha = syllables / count; 
        return alpha;
    }

    public static double countWords(String filename, ArrayList<String> wordArr){
        Scanner infile = new Scanner(System.in);
        try{
            infile = new Scanner(new File(filename));
        }	
        catch (FileNotFoundException ex){
            System.out.println("File Not Found.");
            System.exit(0);
        }	
        double count = 0, alpha = 0;
        String temp;

        while (infile.hasNext()){
            temp = infile.next();               // get next word from file
            wordArr.add(temp);                  // add curr string to array of words
            count++;                            // inc amount of words
        }

        return count;
    }

}

File: cpp/flesch.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cmath>
#include <cstring>
#include <sstream>
#include <iomanip>

using namespace std;

bool isAlphabetic(string check){
    bool alpha = false;
    for (int i = 0; i < check.length(); i++){
        if ((check[i] >= 'a' && check[i] <= 'z') || (check[i] >= 'A' && check[i] <= 'Z')){
            alpha = true;
        }
        else{
            alpha = false;
            break;
        }
    }
    return alpha;
}

bool charIsAlphabetic(char check){
    bool alpha = false;
    if((check >= 'a' && check <= 'z') || (check >= 'A' && check <= 'Z')){
        alpha = true;
    }
    else {
        alpha = false;;
    }
    return alpha;
}

bool isVowel(char check){
    if(check == 'a' || check == 'e' || check == 'i' || check == 'o' || check == 'u' || check == 'y'
            || check == 'A' || check == 'E' || check == 'I' || check == 'O' || check == 'U' || check == 'Y')
        return  true;
    return false;

}

bool isEndofSent(char check){
    if(check == '.' || check == '!' || check == '?' || check == ';' || check == ':')
        return true;
    return false;     
}

double countSyllables(double &sCount, vector<string> &wordArr){
    double syllables = 0;
    for(int j = 0; j < wordArr.size(); j++){ 
        string check = wordArr[j];
        syllables = 0;
        for(int i = 0; i < check.length(); i++){     
            if(check[check.length() - 1] == 'e' && syllables > 1){
                syllables--;
            }
            if(isVowel(check[i]) && !isVowel(check[i+1]) && !isEndofSent(check[i+1])){
                syllables++;
            }
        }
        sCount += syllables;
    }

    return sCount; 
}

int countSentences(string filename, vector<string> sentArr){
    ifstream infile;
    infile.open(filename.c_str());
    double count = 0;
    string temp;
    char ch;
    while(infile.get(ch)){
        if(ch == '.' || ch == '!' || ch == '?' || ch == ';' || ch == ':'){
            count++;
        }
    }
    return count;
}

bool isNum(string str){
    if ((str.find('0') != std::string::npos ||
                str.find('1') != std::string::npos ||
                str.find('2') != std::string::npos ||
                str.find('3') != std::string::npos ||
                str.find('4') != std::string::npos ||
                str.find('5') != std::string::npos ||
                str.find('6') != std::string::npos ||
                str.find('7') != std::string::npos ||
                str.find('8') != std::string::npos ||
                str.find('9') != std::string::npos) &&
            !isAlphabetic(str)){
        return true;
    }

    return false;
}

double getAlpha(string filename, vector<string> wordArr){
    ifstream infile;
    infile.open(filename.c_str());
    double count = 0, alpha = 0, syllables = 0, sCount = 0;
    string temp;
    while(getline(infile, temp, ' ')){
        stringstream ss(temp);
        while(getline(ss, temp, '\n')){
            if(!isNum(temp)){
                wordArr.push_back(temp);
                count++;
            }   
        }
    }
    syllables = countSyllables(sCount, wordArr);
    alpha = syllables / count;
    return alpha;
}

double countWords(string filename, vector<string> wordArr){
    ifstream infile;
    infile.open(filename.c_str());
    double count = 0, alpha = 0;
    string temp;
    while(getline(infile, temp, ' ')){
        if(!isNum(temp)){
            wordArr.push_back(temp);
            count++;
        }
    }
    return count;
}

int main(int argc, char *argv[]){
    double alpha, beta, index, gindex = 0;
    string filename;
    if(argv[1] == NULL){
        cout << "Please enter a file name (must exist in current directory): ";
        cin >> filename;
        argv[1] = new char[filename.length() + 1];
        strcpy(argv[1], filename.c_str()); 
    }
    else{
        filename = argv[1];
    }
    cout << "File being used: " << filename << endl;
    vector <string> wordArr;
    vector <string> sentArr;
    alpha = getAlpha(filename, wordArr);
    beta = countWords(filename, wordArr) / countSentences(filename, sentArr);
    index = 206.835 - alpha*84.6 - beta*1.015;
    gindex = round((alpha*11.8 + beta*0.39 - 15.59) * 10.0) / 10.0;
    cout << "Index = " << round(index) << "\n";
    cout << "Grade Level Index = " << gindex << "\n";
    return 0;
}
